### PR TITLE
#7803 Campaigns back-end

### LINKED ISSUE #7803: Schema migration and back-end for "Campaign" table
@andreievg has already made a start on this -- just need to access his branch and cherry-pick the appropriate commits.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fixes #7803

# 👩🏻‍💻 What does this PR do?

This was pretty much done by @andreievg. I was going to include it with the front-end PR, but I think it'll be needed for other issues in https://github.com/msupply-foundation/open-msupply/issues/7694, so here it is.

Adds "Campaigns" table, plus required database/GraphQL queries and mutations

## 💌 Any notes for the reviewer?

Only way to test for now is using GraphiQL explorer

# 🧪 Testing

- [ ] Test with GraphiQL explorere
- [ ] Wait for front-end to test more thoroughly

# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

### CODE REVIEW

File: server/repository/src/db_diesel/campaign/campaign_row.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+use diesel::prelude::*;
+use serde::{Deserialize, Serialize};
+use chrono::NaiveDate;
+
+table! {
+    campaign(id) {
+        id -> Text,
+        name -> Text,
+        start_date -> Nullable<Date>,
+        end_date -> Nullable<Date>,
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/campaign/campaign_row.rs:14]: I think we should add deleted_datetime and implement soft delete.
We'll still want to associate campaigns with old stock but won't want to see them.
Arguably we can just change the end date to hide it but still think deleted_datetime would be good.

File: server/repository/src/db_diesel/campaign/campaign_row.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        &self,
+        ids: &[String],
+    ) -> Result<Vec<CampaignRow>, RepositoryError> {
+        let result = campaign::table
+            .filter(campaign::id.eq_any(ids))
+            .load(self.connection.lock().connection())?;
+        Ok(result)
+    }
+
+    pub fn delete(&self, campaign_id: &str) -> Result<i64, RepositoryError> {
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/campaign/campaign_row.rs:86]: soft delete?

**Review by jmbrunskill — CHANGES_REQUESTED** [review]
Can you fix postgres tests please (just need to add to get_dates() for postgres I think)
Otherwise looks pretty good, would be keen to implement soft delete but happy to merge as is and come back to that if we need it.

**Comment by CarlosNZ** [comment]
Okay @jmbrunskill, I've added the soft-delete and fixed refresh_dates tests if you want to check it?

Also -- there is sync test for campaigns that Andrei had created, but it wasn't imported and run. I tried including it, but it's giving another error that I don't quite understand. You can see the commented out imports in `/Users/carl/GitHub/open-msupply/server/service/src/sync/test/test_data/mod.rs` if you want to have a look, otherwise I'll leave it till after the front-end API if that's cool?

**Comment by jmbrunskill** [comment]
> Okay @jmbrunskill, I've added the soft-delete and fixed refresh_dates tests if you want to check it?
> 
> Also -- there is sync test for campaigns that Andrei had created, but it wasn't imported and run. I tried including it, but it's giving another error that I don't quite understand. You can see the commented out imports in `/Users/carl/GitHub/open-msupply/server/service/src/sync/test/test_data/mod.rs` if you want to have a look, otherwise I'll leave it till after the front-end API if that's cool?

Pushed a fix, which hopefully works for you too.
It looks like the item_variant and campaign records were using the same ID, Not 100% sure why that is a problem but should be ok now that they're not duplicated.

**Comment by jmbrunskill** [comment]
Also pushed a fix for another test failure due to duplicate test_database names

**Review by jmbrunskill — APPROVED** [review]
Thanks Carl, I think this is good to go now.
